File: workflow/scheduler.py
"""
定时任务服务
使用APScheduler实现定时任务

集成优化后的 core 模块，提供 ETF 数据定时更新功能
"""

from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.cron import CronTrigger
import logging
from django.utils import timezone

logger = logging.getLogger(__name__)

scheduler = None


def get_scheduler():
    """获取调度器实例（单例模式）"""
    global scheduler
    if scheduler is None:
        scheduler = BackgroundScheduler()
        scheduler.start()
        logger.info("定时任务调度器已启动")
    return scheduler


def start_exchange_rate_job():
    """启动汇率更新定时任务"""
    sched = get_scheduler()
    
    # 添加每天10:30执行的任务
    sched.add_job(
        func=update_exchange_rates_job,
        trigger=CronTrigger(hour=10, minute=30),
        id='exchange_rate_update',
        name='汇率更新任务',
        replace_existing=True,
        max_instances=1
    )
    
    logger.info("汇率更新定时任务已配置（每天10:30执行）")


def update_exchange_rates_job():
    """汇率更新任务"""
    from .services_exchange_rate import update_exchange_rates_auto
    from .models import OperationLog

    try:
        logger.info("=" * 60)
        logger.info("开始执行汇率更新任务")
        logger.info("=" * 60)

        # 创建操作记录
        op_log = OperationLog.objects.create(
            operation_type='scheduled_task',
            operation_name='定时汇率更新',
            operator='system',
            status=0  # 进行中
        )

        # 执行更新（从免费API获取真实汇率）
        result = update_exchange_rates_auto()

        # 更新操作记录
        op_log.complete(
            success=True,
            result={'updated_count': result['updated_count']}
        )

        logger.info(f"汇率更新任务执行成功: 更新了 {result['updated_count']} 条记录")
        logger.info("=" * 60)

    except Exception as e:
        logger.error(f"汇率更新任务执行失败: {e}", exc_info=True)

        # 记录失败
        try:
            op_log.complete(
                success=False,
                error=str(e)
            )
        except:
            pass


def update_etf_data(symbols=None):
    """
    更新ETF数据任务
    
    Args:
        symbols: ETF代码列表，None表示更新所有
    
    Returns:
        bool: 是否成功
    """
    from .models import OperationLog
    
    try:
        logger.info("=" * 60)
        logger.info("开始执行ETF数据更新任务")
        logger.info("=" * 60)
        
        # 导入 core 模块的更新功能
        try:
            from core import quick_update
            use_core = True
        except ImportError:
            use_core = False
            logger.warning("core 模块不可用，使用 fallback 更新")
        
        # 创建操作记录
        op_log = OperationLog.objects.create(
            operation_type='scheduled_task',
            operation_name='ETF数据更新',
            operator='system',
            status=0
        )
        
        if use_core:
            # 使用优化的 core 模块
            result = quick_update(symbols=symbols)
            
            success = result['failed_count'] == 0
            message = (
                f"成功: {result['success_count']}, "
                f"失败: {result['failed_count']}, "
                f"跳过: {result['skipped_count']}, "
                f"记录: {result['total_records_updated']}"
            )
        else:
            # Fallback: 使用原有的 services
            from .services import etf_service
            from .models import ETFData
            from datetime import datetime, timedelta
            
            if symbols is None:
                symbols = etf_service.SYMBOLS
            
            success_count = 0
            failed_count = 0
            
            for symbol in symbols:
                try:
                    # 获取历史数据
                    data = etf_service.fetch_historical_data(symbol, '1y')
                    if data is not None and not data.empty:
                        # 保存到数据库
                        for index, row in data.iterrows():
                            ETFData.objects.update_or_create(
                                symbol=symbol,
                                date=index.date(),
                                defaults={
                                    'open_price': float(row.get('Open', 0)),
                                    'high_price': float(row.get('High', 0)),
                                    'low_price': float(row.get('Low', 0)),
                                    'close_price': float(row.get('Close', 0)),
                                    'volume': int(row.get('Volume', 0)),
                                }
                            )
                        success_count += 1
                    else:
                        failed_count += 1
                except Exception as e:
                    logger.error(f"更新 {symbol} 失败: {e}")
                    failed_count += 1
            
            success = failed_count == 0
            message = f"成功: {success_count}, 失败: {failed_count}"
        
        # 更新操作记录
        op_log.complete(
            success=success,
            result={'message': message}
        )
        
        logger.info(f"ETF数据更新完成: {message}")
        logger.info("=" * 60)
        
        return success
        
    except Exception as e:
        logger.error(f"ETF数据更新任务失败: {e}", exc_info=True)
        
        try:
            op_log.complete(success=False, error=str(e))
        except:
            pass
        
        return False


def start_etf_update_job(hour=16, minute=30):
    """
    启动ETF数据定时更新任务
    
    Args:
        hour: 小时 (默认16，美股收盘后)
        minute: 分钟 (默认30)
    """
    sched = get_scheduler()
    
    sched.add_job(
        func=update_etf_data,
        trigger=CronTrigger(hour=hour, minute=minute),
        id='etf_daily_update',
        name=f'ETF每日更新 ({hour:02d}:{minute:02d})',
        replace_existing=True,
        max_instances=1
    )
    
    logger.info(f"ETF数据定时任务已配置（每天{hour:02d}:{minute:02d}执行）")


def start_pre_market_etf_job(hour=9, minute=30):
    """
    启动盘前ETF数据更新任务
    
    Args:
        hour: 小时 (默认9，美股开盘前)
        minute: 分钟 (默认30)
    """
    sched = get_scheduler()
    
    sched.add_job(
        func=update_etf_data,
        trigger=CronTrigger(hour=hour, minute=minute),
        id='etf_pre_market_update',
        name=f'ETF盘前更新 ({hour:02d}:{minute:02d})',
        replace_existing=True,
        max_instances=1
    )
    
    logger.info(f"ETF盘前定时任务已配置（每天{hour:02d}:{minute:02d}执行）")


def start_scheduler():
    """启动调度器并注册所有定时任务"""
    try:
        sched = get_scheduler()
        
        # 启动汇率更新定时任务
        start_exchange_rate_job()
        
        # 启动ETF数据更新任务（收盘后）
        start_etf_update_job(hour=16, minute=30)
        
        # 启动ETF盘前更新任务
        start_pre_market_etf_job(hour=9, minute=30)
        
        logger.info("所有定时任务已启动")
        
    except Exception as e:
        logger.error(f"启动调度器失败: {e}", exc_info=True)


def stop_scheduler():
    """停止调度器"""
    global scheduler
    if scheduler:
        scheduler.shutdown()
        scheduler = None
        logger.info("定时任务调度器已停止")


def list_jobs():
    """列出所有定时任务"""
    sched = get_scheduler()
    jobs = sched.get_jobs()
    
    job_list = []
    for job in jobs:
        job_list.append({
            'id': job.id,
            'name': job.name,
            'next_run_time': job.next_run_time,
            'trigger': str(job.trigger)
        })
    
    return job_list

File: backend/tasks/scheduler.go
package tasks

import (
	"context"
	"time"

	"etf-insight/config"
	"etf-insight/models"
	"etf-insight/services"
	"etf-insight/services/datasource"
	"etf-insight/utils"

	"github.com/robfig/cron/v3"
	"github.com/shopspring/decimal"
	"gorm.io/gorm/clause"
)

// Scheduler 定时任务调度器
type Scheduler struct {
	cron        *cron.Cron
	cfg         *config.ScheduleConfig
	analysisSvc *services.ETFAnalysisService
	exchangeSvc *services.ExchangeRateService
	provider    datasource.DataSourceProvider
}

// NewScheduler 创建新的调度器
func NewScheduler(cfg *config.ScheduleConfig, analysis *services.ETFAnalysisService, exchange *services.ExchangeRateService, provider datasource.DataSourceProvider) *Scheduler {
	c := cron.New(cron.WithSeconds())

	return &Scheduler{
		cron:        c,
		cfg:         cfg,
		analysisSvc: analysis,
		exchangeSvc: exchange,
		provider:    provider,
	}
}

// Start 启动调度器
func (s *Scheduler) Start() {
	utils.Info("Starting scheduler...")

	// 添加汇率更新任务 (每天 10:30)
	_, err := s.cron.AddFunc("0 30 10 * * *", s.updateExchangeRates)
	if err != nil {
		utils.Error("Failed to add exchange rate update job", err)
	} else {
		utils.Info("Exchange rate update job scheduled at 10:30 daily")
	}

	// 添加ETF盘前更新任务 (每天 9:30)
	_, err = s.cron.AddFunc("0 30 9 * * *", s.updateETFData)
	if err != nil {
		utils.Error("Failed to add ETF pre-market update job", err)
	} else {
		utils.Info("ETF pre-market update job scheduled at 09:30 daily")
	}

	// 添加ETF收盘后更新任务 (每天 16:30)
	_, err = s.cron.AddFunc("0 30 16 * * *", s.updateETFData)
	if err != nil {
		utils.Error("Failed to add ETF post-market update job", err)
	} else {
		utils.Info("ETF post-market update job scheduled at 16:30 daily")
	}

	// 添加每小时检查任务
	_, err = s.cron.AddFunc("0 0 * * * *", s.hourlyCheck)
	if err != nil {
		utils.Error("Failed to add hourly check job", err)
	} else {
		utils.Info("Hourly check job scheduled")
	}

	s.cron.Start()
	utils.Info("Scheduler started successfully")
}

// Stop 停止调度器
func (s *Scheduler) Stop() {
	utils.Info("Stopping scheduler...")
	ctx := s.cron.Stop()
	<-ctx.Done()
	utils.Info("Scheduler stopped")
}

// updateExchangeRates 更新汇率
func (s *Scheduler) updateExchangeRates() {
	utils.Info("Running scheduled exchange rate update...")

	opLog := models.OperationLog{
		OperationType: "scheduled_task",
		OperationName: "定时汇率更新",
		Operator:      "system",
		Status:        0,
		StartTime:     time.Now(),
	}
	models.DB.Create(&opLog)

	if s.exchangeSvc != nil {
		if err := s.exchangeSvc.UpdateRates(); err != nil {
			utils.Error("Failed to update exchange rates", err)
			opLog.Status = 2
			opLog.ErrorMessage = err.Error()
		} else {
			utils.Info("Exchange rates updated successfully")
			opLog.Status = 1
		}
	}

	endTime := time.Now()
	opLog.EndTime = &endTime
	opLog.DurationMs = int(endTime.Sub(opLog.StartTime).Milliseconds())
	models.DB.Save(&opLog)
}

// updateETFData 更新ETF数据 - 使用Finage聚合API获取完整OHLCV数据并入库
func (s *Scheduler) updateETFData() {
	utils.Info("Running scheduled ETF data update...")

	opLog := models.OperationLog{
		OperationType: "scheduled_task",
		OperationName: "ETF数据更新",
		Operator:      "system",
		Status:        0,
		StartTime:     time.Now(),
	}
	models.DB.Create(&opLog)

	// 获取所有启用的ETF配置（从数据库读取，不硬编码）
	var etfConfigs []models.ETFConfig
	if err := models.DB.Where("status = ?", 1).Find(&etfConfigs).Error; err != nil || len(etfConfigs) == 0 {
		utils.Warn("No enabled ETF configs found", "error", err)
		opLog.Status = 2
		opLog.ErrorMessage = "no enabled ETF configs"
		endTime := time.Now()
		opLog.EndTime = &endTime
		models.DB.Save(&opLog)
		return
	}

	symbols := make([]string, 0, len(etfConfigs))
	for _, cfg := range etfConfigs {
		symbols = append(symbols, cfg.Symbol)
	}

	// 使用数据源获取完整OHLCV数据
	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()

	if s.provider == nil || !s.provider.IsAvailable(ctx) {
		providerName := "nil"
		if s.provider != nil {
			providerName = s.provider.GetName()
		}
		utils.Warn("Data source not available", "provider", providerName)
		opLog.Status = 2
		opLog.ErrorMessage = "data source not available: " + providerName
		endTime := time.Now()
		opLog.EndTime = &endTime
		models.DB.Save(&opLog)
		return
	}

	// 逐个获取ETF数据（使用聚合API获取完整OHLCV）
	quotes, err := s.provider.GetQuotes(ctx, symbols)
	if err != nil {
		utils.Error("Failed to get quotes from data source", err, "provider", s.provider.GetName())
		opLog.Status = 2
		opLog.ErrorMessage = err.Error()
		endTime := time.Now()
		opLog.EndTime = &endTime
		models.DB.Save(&opLog)
		return
	}

	// 入库完整OHLCV数据
	successCount := 0
	failCount := 0
	for _, quote := range quotes {
		// 检查OHLCV数据是否完整，不完整的跳过不入库
		if quote.OpenPrice == 0 && quote.DayHigh == 0 && quote.DayLow == 0 && quote.Volume == 0 {
			utils.Warn("Skipping incomplete OHLCV data",
				"symbol", quote.Symbol,
				"dataSource", quote.DataSource,
				"currentPrice", quote.CurrentPrice)
			failCount++
			continue
		}

		// 确定日期：使用聚合API返回的Timestamp
		date := quote.Timestamp
		if date.IsZero() {
			date = time.Now().Truncate(24 * time.Hour)
		}

		// Upsert 数据 (symbol + date 联合唯一)
		etfData := models.ETFData{
			Symbol:     quote.Symbol,
			Date:       date,
			OpenPrice:  decimal.NewFromFloat(quote.OpenPrice),
			ClosePrice: decimal.NewFromFloat(quote.CurrentPrice),
			HighPrice:  decimal.NewFromFloat(quote.DayHigh),
			LowPrice:   decimal.NewFromFloat(quote.DayLow),
			Volume:     quote.Volume,
			DataSource: quote.DataSource,
		}

		result := models.DB.Clauses(clause.OnConflict{
			Columns:   []clause.Column{{Name: "symbol"}, {Name: "date"}},
			DoUpdates: clause.AssignmentColumns([]string{"open_price", "close_price", "high_price", "low_price", "volume", "data_source"}),
		}).Create(&etfData)

		if result.Error != nil {
			utils.Warn("Failed to upsert ETF data", result.Error, "symbol", quote.Symbol)
			failCount++
			continue
		}

		successCount++
		utils.Info("Updated ETF data",
			"symbol", quote.Symbol,
			"open", quote.OpenPrice,
			"high", quote.DayHigh,
			"low", quote.DayLow,
			"close", quote.CurrentPrice,
			"volume", quote.Volume,
			"source", quote.DataSource)

		// 缓存已移除，不再更新缓存中的实时数据
	}

	utils.Info("ETF data update completed", "success", successCount, "fail", failCount, "total", len(symbols), "provider", s.provider.GetName())

	opLog.Status = 1
	if failCount > 0 && successCount == 0 {
		opLog.Status = 2
	} else if failCount > 0 {
		opLog.Status = 3
	}
	endTime := time.Now()
	opLog.EndTime = &endTime
	opLog.DurationMs = int(endTime.Sub(opLog.StartTime).Milliseconds())
	models.DB.Save(&opLog)
}

// hourlyCheck 每小时检查
func (s *Scheduler) hourlyCheck() {
	utils.Info("Running hourly check...")

	// 缓存已移除：if s.cacheService != nil {
	//	stats := s.cacheService.GetCacheStats()
	//	utils.Info("Cache stats", "stats", stats)
	// }
}

// RunOnce 立即执行一次更新
func (s *Scheduler) RunOnce() {
	utils.Info("Running one-time update...")
	s.updateETFData()
	s.updateExchangeRates()
}

// GetJobs 获取所有任务
func (s *Scheduler) GetJobs() []map[string]interface{} {
	entries := s.cron.Entries()
	var jobs []map[string]interface{}

	for _, entry := range entries {
		jobs = append(jobs, map[string]interface{}{
			"id":       entry.ID,
			"next_run": entry.Next.Format("2006-01-02 15:04:05"),
			"prev_run": entry.Prev.Format("2006-01-02 15:04:05"),
		})
	}

	return jobs
}
